scrapers/zilean.py

```diff
@@ -0,0 +1,217 @@
+import logging
+import re
+from datetime import datetime, timedelta
+
+import PTN
+import httpx
+from pymongo.errors import DuplicateKeyError
+from redis.asyncio import Redis
+
+from db.config import settings
+from db.models import TorrentStreams, Season, Episode
+from db.schemas import TorrentStreamsList
+from scrapers.helpers import (
+    update_torrent_series_streams_metadata,
+    update_torrent_movie_streams_metadata,
+)
+from utils.const import UA_HEADER
+from utils.parser import (
+    convert_size_to_bytes,
+    is_contain_18_plus_keywords,
+)
+
+
+async def get_streams_from_zilean(
+    redis: Redis,
+    streams: list[TorrentStreams],
+    video_id: str,
+    catalog_type: str,
+    title: str,
+    season: int = None,
+    episode: int = None,
+):
+    cache_key = f"{catalog_type}_{video_id}_{season}_{episode}_zilean_dmm_streams"
+    cached_data = await redis.get(cache_key)
+    if cached_data:
+        return TorrentStreamsList.model_validate_json(cached_data).streams
+
+    if catalog_type == "movie":
+        streams.extend(
+            await scrap_movie_streams_from_zilean(
+                video_id, title
+            )
+        )
+    elif catalog_type == "series":
+        streams.extend(
+            await scrap_series_streams_from_zilean(
+                video_id, title, season, episode
+            )
+        )
+
+    # Cache the data for 24 hours
+    await redis.set(
+        cache_key,
+        TorrentStreamsList(streams=streams).model_dump_json(exclude_none=True),
+        ex=int(timedelta(hours=settings.prowlarr_search_interval_hour).total_seconds()),
+    )
+
+    return streams
+
+
+async def fetch_stream_data(url: str) -> list:
+    """Fetch stream data asynchronously."""
+    async with httpx.AsyncClient(
+        headers=UA_HEADER, proxy=settings.scraper_proxy_url
+    ) as client:
+        response = await client.get(url, timeout=10)
+        response.raise_for_status()  # Will raise an exception for 4xx/5xx responses
+        return response.json()
+
+
+async def scrap_movie_streams_from_zilean(
+    video_id: str, title: str,
+) -> list[TorrentStreams]:
+    """
+    Get streams by text from zilean DMM.
+    """
+    url = f"{settings.zilean_url}/dmm/filtered?Query={title}"
```

Comment:
zilean using parsett library to parse the data. I'm also going to use the same. However that implementation is bit old and im adding new parsing capability. 

zilean also support raw search without filtering with matching queries. 

Ex: 
```
curl --location 'http://zilean.zilean:8181/dmm/search' \
--header 'Content-Type: application/json' \
--data '{"queryText": "Hit Man"}'
```

Whats your thought? do we use already parsed data or get the data and parse it from our side?

Reply:
We can parse it from the raw title if you want and then use the parsett library to parse the data. We can use the same endpoint I am using here for now and that has the raw title anyways, and you can add the parsett parsing once this is checked in and replace how you get the info?

So anyways this endpoint provides more info and can be used to enhance whatever we need?